statistics/selectivity.go

```diff
@@ -0,0 +1,221 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package statistics
+
+import (
+	"math"
+
+	"github.com/juju/errors"
+	"github.com/pingcap/tidb/context"
+	"github.com/pingcap/tidb/expression"
+	"github.com/pingcap/tidb/sessionctx/variable"
+	"github.com/pingcap/tidb/util/ranger"
+	"github.com/pingcap/tidb/util/types"
+)
+
+// If one condition can't be calculated, we will assume that the selectivity of this condition is 0.8.
+const selectionFactor = 0.8
+
+// calcBlock is used for calculating selectivity.
+type calcBlock struct {
+	tp int
+	ID int64
+	// the ith bit of `cover` will tell whether the ith expression is covered by this index.
+	cover int64
+	// ranges calculated.
+	ranges []types.Range
+}
+
+// The type of the SelUnit if it isn't a index.
+const (
+	SelPkUnit     = -1
+	SelColumnUnit = -2
+)
+
+// SelUnit is used as a param passed by Selectivity.
+// When ID >= 0, this unit is a index. And the ID, cols, lengths is the information of this index.
+// When ID is SelPkUnit, cols' len is 0, and it stores the pk column.
+// When ID is SelColumnUnit, cols stores the columns used for column sampling calculation.
+type SelUnit struct {
+	ID      int64
+	cols    []*expression.Column
+	lengths []int
+}
+
+// MakeSelUnit generates a SelUnit by the params.
+func MakeSelUnit(id int64, lengths []int, cols ...*expression.Column) *SelUnit {
+	return &SelUnit{ID: id, cols: cols, lengths: lengths}
+}
+
+// Selectivity is a function calculate the selectivity of the expressions.
+// The definition of selectivity is (row count after filter / row count before filter).
+// And exprs must be CNF now, in other words, `exprs[0] and exprs[1] and ... and exprs[len - 1]` should be held when you call this.
+// TODO: support expressions that the top layer is a DNF.
+// Currently the time complexity if o(n^2).
+func Selectivity(ctx context.Context, exprs []expression.Expression, units []*SelUnit, statsTbl *Table) (float64, error) {
+	// TODO: If len(exprs) is bigger than 63, we could use bitset structure to replace the int64.
+	// This will simplify some code and speed up if we use this rather than a boolean slice.
+	if statsTbl.Pseudo || len(exprs) > 63 {
+		return selectionFactor, nil
+	}
+	var blocks []*calcBlock
+	sc := ctx.GetSessionVars().StmtCtx
+	for _, unit := range units {
+		switch unit.ID {
+		case SelPkUnit:
```

Comment:
Below there will need to call `Ranges2IntRanges` or `Ranges2ColumnRange`, and  `GetRowCountByIntColumnRanges` or `GetRowCountByColumnRanges`。